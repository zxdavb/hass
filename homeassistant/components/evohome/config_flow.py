"""Config flow to configure the evohome integration."""

from datetime import timedelta
from functools import partial
import logging
from typing import Any, Final

import evohomeasync2 as evo
import voluptuous as vol  # type: ignore[import-untyped, unused-ignore]

from homeassistant import config_entries as ce
from homeassistant.config_entries import ConfigEntry, ConfigFlowResult, OptionsFlow
from homeassistant.const import CONF_PASSWORD, CONF_SCAN_INTERVAL, CONF_USERNAME
from homeassistant.core import callback
from homeassistant.helpers.aiohttp_client import async_get_clientsession
import homeassistant.helpers.config_validation as cv

from .const import DOMAIN
from .helpers import handle_exception

CONF_LOCATION_IDX: Final = "location_idx"

SCAN_INTERVAL_DEFAULT: Final = timedelta(seconds=300)
SCAN_INTERVAL_MINIMUM: Final = timedelta(seconds=60)

_CONFIG_KEYS: Final = (CONF_USERNAME, CONF_PASSWORD, CONF_LOCATION_IDX)

_LOGGER = logging.getLogger(__name__)


class EvoConfigFlow(ce.ConfigFlow, domain=DOMAIN):
    """Handle an evohome config flow."""

    VERSION = 1
    MINOR_VERSION = 1

    def __init__(self) -> None:
        """Initialize the config flow."""

        _LOGGER.error(f"{self}: __init__()")  # noqa: G004

        self.options: dict[str, Any] = {}  # TO-DO: or: Mapping[str, Any]
        self.username: str | None = None
        self.password: str | None = None
        self.client: evo.EvohomeClient | None = None

    async def async_step_import(self, import_data: dict[str, Any]) -> ConfigFlowResult:
        """Import a config_entry from configuration.yaml."""

        _LOGGER.warning(f"{self}: async_step_import({import_data})")  # noqa: G004

        data = {k: v for k, v in import_data.items() if k in _CONFIG_KEYS}
        options = {k: v for k, v in import_data.items() if k not in _CONFIG_KEYS}

        return self.async_create_entry(title=DOMAIN, data=data, options=options)

    async def async_step_user(
        self, user_input: dict[str, Any] | None = None
    ) -> ConfigFlowResult:
        """Handle a flow initialized by the user."""

        _LOGGER.error(f"{self}: async_step_user({user_input})")  # noqa: G004

        return await self.async_step_user_credentials(user_input)

    async def async_step_user_credentials(
        self, user_input: dict[str, Any] | None = None
    ) -> ConfigFlowResult:
        """Handle the user credentials step."""

        _LOGGER.error(f"{self}: async_step_user_credentials({user_input})")  # noqa: G004

        errors = {}

        if user_input is not None:  # TO-DO: validate the user input
            partial_fnc = partial(
                evo.EvohomeClient,
                user_input[CONF_USERNAME],
                user_input[CONF_PASSWORD],
                # **tokens,
                session=async_get_clientsession(self.hass),
            )
            try:
                client = await self.hass.async_add_executor_job(partial_fnc)
            except evo.AuthenticationFailed as err:
                handle_exception(err)
                errors["base"] = "invalid_auth"
            else:
                # username/password valid so show user locations
                self.username = user_input[CONF_USERNAME]
                self.password = user_input[CONF_PASSWORD]
                self.client = client

                return await self.async_step_location_idx(user_input)

        return self.async_show_form(
            step_id="user_credentials",
            data_schema=vol.Schema(
                {
                    vol.Required(CONF_USERNAME): cv.string,
                    vol.Required(CONF_PASSWORD): cv.string,
                }
            ),
            errors=errors,
        )

    async def async_step_location_idx(
        self, user_input: dict[str, Any] | None = None
    ) -> ConfigFlowResult:
        """Handle the location index step."""

        _LOGGER.warning(f"{self}: async_step_location_idx({user_input})")  # noqa: G004

        if user_input is not None:  # TO-DO: validate the user input
            return self.async_create_entry(title=DOMAIN, data=user_input)

        return self.async_show_form(
            step_id=CONF_LOCATION_IDX,
            data_schema=vol.Schema(
                {
                    vol.Optional(CONF_LOCATION_IDX, default=0): cv.positive_int,
                }
            ),
        )

    @staticmethod
    @callback
    def async_get_options_flow(config_entry: ConfigEntry) -> OptionsFlow:
        """Get the options flow for this handler."""

        _LOGGER.warning("EvoConfigFlow: async_get_options_flow()")  # noqa: G004

        return EvoOptionsFlow(config_entry)


class EvoOptionsFlow(OptionsFlow):
    """Config flow options for Evohome."""

    def __init__(self, config_entry: ConfigEntry) -> None:
        """Initialize Evohome options flow."""

        _LOGGER.warning(f"{self}: __init__({config_entry})")  # noqa: G004

        self.config_entry = config_entry

    async def async_step_init(
        self, user_input: dict[str, Any] | None = None
    ) -> ConfigFlowResult:
        """Manage the options."""

        _LOGGER.warning(f"{self}: async_step_init({user_input})")  # noqa: G004

        if user_input is not None:
            return self.async_create_entry(title=DOMAIN, data=user_input)

        options = self.config_entry.options
        return self.async_show_form(
            step_id="init",
            data_schema=vol.Schema(
                {
                    vol.Required(
                        CONF_SCAN_INTERVAL,
                        default=options.get(CONF_SCAN_INTERVAL, SCAN_INTERVAL_DEFAULT),
                    ): vol.All(cv.time_period, vol.Range(min=SCAN_INTERVAL_MINIMUM)),
                }
            ),
        )
