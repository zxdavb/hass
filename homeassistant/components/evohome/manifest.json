{
  "domain": "evohome",
  "name": "Honeywell Total Connect Comfort (Europe)",
  "codeowners": ["@zxdavb"],
  "config_flow": true,
  "documentation": "https://www.home-assistant.io/integrations/evohome",
  "iot_class": "cloud_polling",
  "loggers": ["evohomeasync", "evohomeasync2"],
  "requirements": ["evohome-async==0.4.19"],
  "single_config_entry": true
}
