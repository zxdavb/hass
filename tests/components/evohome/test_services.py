"""The tests for evohome services.

Evohome implements some domain-specific service.
"""

from __future__ import annotations

from unittest.mock import patch

from freezegun.api import FrozenDateTimeFactory
import pytest
from syrupy import SnapshotAssertion

from homeassistant.components.climate import (
    ATTR_HVAC_MODE,
    ATTR_PRESET_MODE,
    ATTR_TEMPERATURE,
    SERVICE_SET_HVAC_MODE,
    SERVICE_SET_PRESET_MODE,
    SERVICE_SET_TEMPERATURE,
    HVACMode,
)
from homeassistant.components.evohome import DOMAIN, EvoService
from homeassistant.components.water_heater import (
    ATTR_AWAY_MODE,
    ATTR_OPERATION_MODE,
    SERVICE_SET_AWAY_MODE,
    SERVICE_SET_OPERATION_MODE,
)
from homeassistant.const import (
    ATTR_ENTITY_ID,
    SERVICE_TURN_OFF,
    SERVICE_TURN_ON,
    Platform,
)
from homeassistant.core import HomeAssistant
from homeassistant.exceptions import HomeAssistantError, ServiceValidationError
from homeassistant.util import dt as dt_util

from .conftest import ctl_entity, dhw_entity, setup_evohome, zone_entity
from .const import CTL_MODE_LOOKUP, TEST_INSTALLS


@pytest.mark.parametrize("install", TEST_INSTALLS)
async def test_evohome_ctl_svcs(
    hass: HomeAssistant,
    config: dict[str, str],
    install: str,
    freezer: FrozenDateTimeFactory,
    snapshot: SnapshotAssertion,
) -> None:
    """Test domain-specific services of a evohome-compatible controller."""

    freezer.move_to("2024-07-10T12:00:00Z")
    results = []

    async for _ in setup_evohome(hass, config, install=install):
        ctl = ctl_entity(hass)

        services = list(hass.services.async_services_for_domain(DOMAIN))
        assert services == snapshot

        # EvoService.REFRESH_SYSTEM
        with patch("evohomeasync2.location.Location.refresh_status") as mock_fcn:
            await hass.services.async_call(
                DOMAIN,
                EvoService.REFRESH_SYSTEM,
                service_data=None,
                blocking=True,
            )

            assert mock_fcn.await_count == 1
            assert mock_fcn.await_args.args == ()
            assert mock_fcn.await_args.kwargs == {}

        # EvoService.SET_SYSTEM_MODE: HeatingOff (or Off)
        ctl_mode = "Off" if "Off" in ctl._modes else "HeatingOff"  # most are HeatingOff

        with patch("evohomeasync2.controlsystem.ControlSystem.set_mode") as mock_fcn:
            await hass.services.async_call(
                DOMAIN,
                EvoService.SET_SYSTEM_MODE,
                service_data={"mode": ctl_mode},
                blocking=True,
            )

            assert mock_fcn.await_count == 1
            assert mock_fcn.await_args.args == (ctl_mode,)
            assert mock_fcn.await_args.kwargs == {
                "until": None,
            }

        # EvoService.SET_SYSTEM_MODE: Auto (or Heat)
        ctl_mode = "Auto" if "Auto" in ctl._modes else "Heat"  # most are Auto

        with patch("evohomeasync2.controlsystem.ControlSystem.set_mode") as mock_fcn:
            await hass.services.async_call(
                DOMAIN,
                EvoService.SET_SYSTEM_MODE,
                service_data={"mode": ctl_mode},
                blocking=True,
            )

            assert mock_fcn.await_count == 1
            assert mock_fcn.await_args.args == (ctl_mode,)
            assert mock_fcn.await_args.kwargs == {
                "until": None,
            }

        # some systems (e.g. VisionPro Wifi) support only Off, Heat
        if not [
            m for m in ctl._modes if m not in ("Auto", "Heat", "HeatingOff", "Off")
        ]:
            return

        # EvoService.SET_SYSTEM_MODE: Away, for 21 days
        with patch("evohomeasync2.controlsystem.ControlSystem.set_mode") as mock_fcn:
            await hass.services.async_call(
                DOMAIN,
                EvoService.SET_SYSTEM_MODE,
                service_data={"mode": "Away", "period": {"days": 21}},
                blocking=True,
            )

            assert mock_fcn.await_count == 1
            assert mock_fcn.await_args.args == ("Away",)
            assert install != "default" or mock_fcn.await_args.kwargs == {
                "until": dt_util.parse_datetime("2024-07-31T07:00:00Z")
            }

            results.append(mock_fcn.await_args.kwargs)

        # EvoService.SET_SYSTEM_MODE: AutoWithEco, for 150 minutes
        with patch("evohomeasync2.controlsystem.ControlSystem.set_mode") as mock_fcn:
            await hass.services.async_call(
                DOMAIN,
                EvoService.SET_SYSTEM_MODE,
                service_data={"mode": "AutoWithEco", "duration": {"minutes": 150}},
                blocking=True,
            )

            assert mock_fcn.await_count == 1
            assert mock_fcn.await_args.args == ("AutoWithEco",)
            assert mock_fcn.await_args.kwargs == {
                "until": dt_util.parse_datetime("2024-07-10T14:30:00Z")
            }

        # only evohome (e.g. not RoundThermostat) supports AutoWithReset
        if "AutoWithReset" not in ctl._modes:
            # not need to confirm EvoService.RESET_SYSTEM not in services
            return

        # EvoService.RESET_SYSTEM
        with patch("evohomeasync2.controlsystem.ControlSystem.set_mode") as mock_fcn:
            await hass.services.async_call(
                DOMAIN,
                EvoService.RESET_SYSTEM,
                service_data=None,
                blocking=True,
            )

            assert mock_fcn.await_count == 1
            assert mock_fcn.await_args.args == ("AutoWithReset",)
            assert mock_fcn.await_args.kwargs == {"until": None}

    assert results == snapshot


@pytest.mark.parametrize("install", TEST_INSTALLS)
async def test_evohome_zon_svcs(
    hass: HomeAssistant,
    config: dict[str, str],
    install: str,
    freezer: FrozenDateTimeFactory,
    snapshot: SnapshotAssertion,
) -> None:
    """Test domain-specific services of a evohome-compatible climate zone."""

    freezer.move_to("2024-07-10T12:00:00Z")
    results = []

    async for _ in setup_evohome(hass, config, install=install):
        zone = zone_entity(hass)

        # reset zone to follow_schedule
        with patch("evohomeasync2.zone.Zone.reset_mode") as mock_fcn:
            await hass.services.async_call(
                DOMAIN,
                EvoService.RESET_ZONE_OVERRIDE,
                service_data={"entity_id": zone.entity_id},
                blocking=True,
            )

            assert mock_fcn.await_count == 1
            assert mock_fcn.await_args.args == ()
            assert mock_fcn.await_args.kwargs == {}

        # set zone to temporary_override
        with patch("evohomeasync2.zone.Zone.set_temperature") as mock_fcn:
            await hass.services.async_call(
                DOMAIN,
                EvoService.SET_ZONE_OVERRIDE,
                service_data={
                    "entity_id": zone.entity_id,
                    "setpoint": 20.1,
                    "duration": {"minutes": 150},
                },
                blocking=True,
            )

            assert mock_fcn.await_count == 1
            assert mock_fcn.await_args.args == (20.1,)
            assert mock_fcn.await_args.kwargs == {
                "until": dt_util.parse_datetime("2024-07-10T14:30:00Z")
            }

        # set zone to temporary_override (advance to next scheduled setpoint)
        with patch("evohomeasync2.zone.Zone.set_temperature") as mock_fcn:
            await hass.services.async_call(
                DOMAIN,
                EvoService.SET_ZONE_OVERRIDE,
                service_data={
                    "entity_id": zone.entity_id,
                    "setpoint": 20.2,
                    "duration": 0,
                },
                blocking=True,
            )

            assert mock_fcn.await_count == 1
            assert mock_fcn.await_args.args == (20.2,)
            assert install != "default" or mock_fcn.await_args.kwargs == {
                "until": dt_util.parse_datetime("2024-07-10T21:10:00Z"),
            }

            results.append(mock_fcn.await_args.kwargs)

        # set zone to permanent_override
        with patch("evohomeasync2.zone.Zone.set_temperature") as mock_fcn:
            await hass.services.async_call(
                DOMAIN,
                EvoService.SET_ZONE_OVERRIDE,
                service_data={
                    "entity_id": zone.entity_id,
                    "setpoint": 20.3,
                },
                blocking=True,
            )

        assert mock_fcn.await_count == 1
        assert mock_fcn.await_args.args == (20.3,)
        assert mock_fcn.await_args.kwargs == {
            "until": None,
        }

    assert results == snapshot


@pytest.mark.parametrize("install", TEST_INSTALLS)
async def test_climate_ctl_svcs(
    hass: HomeAssistant,
    config: dict[str, str],
    install: str,
    freezer: FrozenDateTimeFactory,
    snapshot: SnapshotAssertion,
) -> None:
    """Test climate services of a evohome-compatible controller."""

    freezer.move_to("2024-07-10T12:00:00Z")
    results = []

    async for _ in setup_evohome(hass, config, install=install):
        ctl = ctl_entity(hass)

        assert ctl.hvac_modes == [HVACMode.OFF, HVACMode.HEAT]

        # SERVICE_SET_HVAC_MODE: HVACMode.OFF
        ctl_mode = "HeatingOff" if "HeatingOff" in ctl._modes else "Off"

        with patch("evohomeasync2.controlsystem.ControlSystem.set_mode") as mock_fcn:
            await hass.services.async_call(
                Platform.CLIMATE,
                SERVICE_SET_HVAC_MODE,
                {
                    ATTR_ENTITY_ID: ctl.entity_id,
                    ATTR_HVAC_MODE: HVACMode.OFF,
                },
                blocking=True,
            )

            assert mock_fcn.await_count == 1
            assert mock_fcn.await_args.args == (ctl_mode,)
            assert mock_fcn.await_args.kwargs == {"until": None}

        # SERVICE_SET_HVAC_MODE: HVACMode.HEAT
        ctl_mode = "Heat" if "Heat" in ctl._modes else "Auto"

        with patch("evohomeasync2.controlsystem.ControlSystem.set_mode") as mock_fcn:
            await hass.services.async_call(
                Platform.CLIMATE,
                SERVICE_SET_HVAC_MODE,
                {
                    ATTR_ENTITY_ID: ctl.entity_id,
                    ATTR_HVAC_MODE: HVACMode.HEAT,
                },
                blocking=True,
            )

            assert mock_fcn.await_count == 1
            assert mock_fcn.await_args.args == (ctl_mode,)
            assert mock_fcn.await_args.kwargs == {"until": None}

        assert install != "default" or ctl.preset_modes == list(CTL_MODE_LOOKUP)
        assert ctl.preset_modes == snapshot

        # SERVICE_SET_PRESET_MODE: various
        for mode in ctl.preset_modes or []:  # could be None
            with patch(
                "evohomeasync2.controlsystem.ControlSystem.set_mode"
            ) as mock_fcn:
                await hass.services.async_call(
                    Platform.CLIMATE,
                    SERVICE_SET_PRESET_MODE,
                    {
                        ATTR_ENTITY_ID: ctl.entity_id,
                        ATTR_PRESET_MODE: mode,
                    },
                    blocking=True,
                )

                assert mock_fcn.await_count == 1
                assert install != "default" or mock_fcn.await_args.args == (
                    CTL_MODE_LOOKUP[mode],
                )
                assert mock_fcn.await_args.kwargs == {"until": None}

                results.append(mock_fcn.await_args.args)

        # SERVICE_SET_TEMPERATURE (not supported)
        with pytest.raises(HomeAssistantError):
            await hass.services.async_call(
                Platform.CLIMATE,
                SERVICE_SET_TEMPERATURE,
                {
                    ATTR_ENTITY_ID: ctl.entity_id,
                    ATTR_TEMPERATURE: 23.0,
                },
                blocking=True,
            )

    assert results == snapshot


@pytest.mark.parametrize("install", TEST_INSTALLS)
async def test_climate_zon_svcs(
    hass: HomeAssistant,
    config: dict[str, str],
    install: str,
    freezer: FrozenDateTimeFactory,
    snapshot: SnapshotAssertion,
) -> None:
    """Test climate services of a evohome-compatible climate zone."""

    freezer.move_to("2024-07-10T12:00:00Z")
    results = []

    async for _ in setup_evohome(hass, config, install=install):
        zone = zone_entity(hass)

        assert zone.preset_modes == ["none", "temporary", "permanent"]

        # SERVICE_SET_PRESET_MODE: "none"
        with patch("evohomeasync2.zone.Zone.reset_mode") as mock_fcn:
            await hass.services.async_call(
                Platform.CLIMATE,
                SERVICE_SET_PRESET_MODE,
                {
                    ATTR_ENTITY_ID: zone.entity_id,
                    ATTR_PRESET_MODE: "none",
                },
                blocking=True,
            )

            assert mock_fcn.await_count == 1
            assert mock_fcn.await_args.args == ()
            assert mock_fcn.await_args.kwargs == {}

        # SERVICE_SET_PRESET_MODE: "permanent"
        with patch("evohomeasync2.zone.Zone.set_temperature") as mock_fcn:
            await hass.services.async_call(
                Platform.CLIMATE,
                SERVICE_SET_PRESET_MODE,
                {
                    ATTR_ENTITY_ID: zone.entity_id,
                    ATTR_PRESET_MODE: "permanent",
                },
                blocking=True,
            )

            assert mock_fcn.await_count == 1
            assert install != "default" or mock_fcn.await_args.args == (17.0,)
            assert mock_fcn.await_args.kwargs == {
                "until": None,
            }

            results.append(mock_fcn.await_args.args)

        # SERVICE_SET_PRESET_MODE: "temporary"
        with patch("evohomeasync2.zone.Zone.set_temperature") as mock_fcn:
            await hass.services.async_call(
                Platform.CLIMATE,
                SERVICE_SET_PRESET_MODE,
                {
                    ATTR_ENTITY_ID: zone.entity_id,
                    ATTR_PRESET_MODE: "temporary",
                },
                blocking=True,
            )

            assert mock_fcn.await_count == 1
            assert install != "default" or mock_fcn.await_args.args == (17.0,)
            assert install != "default" or mock_fcn.await_args.kwargs == {
                "until": dt_util.parse_datetime("2024-07-10T21:10:00Z")
            }

            results.append(mock_fcn.await_args)

        # SERVICE_SET_TEMPERATURE: 23.0
        with patch("evohomeasync2.zone.Zone.set_temperature") as mock_fcn:
            await hass.services.async_call(
                Platform.CLIMATE,
                SERVICE_SET_TEMPERATURE,
                {
                    ATTR_ENTITY_ID: zone.entity_id,
                    ATTR_TEMPERATURE: 23.0,
                },
                blocking=True,
            )

            assert mock_fcn.await_count == 1
            assert mock_fcn.await_args.args == (23.0,)
            assert "until" in mock_fcn.await_args.kwargs  # to next scheduled setpoint

            results.append(mock_fcn.await_args.kwargs)  # varies by install fixture/zone

        for temperature in (0, 40):  # outside of min/max temp range
            with pytest.raises(ServiceValidationError):
                await hass.services.async_call(
                    Platform.CLIMATE,
                    SERVICE_SET_TEMPERATURE,
                    {
                        ATTR_ENTITY_ID: zone.entity_id,
                        ATTR_TEMPERATURE: temperature,
                    },
                    blocking=True,
                )

        assert zone.hvac_modes == [HVACMode.OFF, HVACMode.HEAT]

        # SERVICE_SET_HVAC_MODE: HVACMode.OFF
        with patch("evohomeasync2.zone.Zone.set_temperature") as mock_fcn:
            await hass.services.async_call(
                Platform.CLIMATE,
                SERVICE_SET_HVAC_MODE,
                {
                    ATTR_ENTITY_ID: zone.entity_id,
                    ATTR_HVAC_MODE: HVACMode.OFF,
                },
                blocking=True,
            )

            assert mock_fcn.await_count == 1
            assert mock_fcn.await_args.args == (zone.min_temp,)
            assert mock_fcn.await_args.kwargs == {"until": None}

        # SERVICE_SET_HVAC_MODE: HVACMode.HEAT
        with patch("evohomeasync2.zone.Zone.reset_mode") as mock_fcn:
            await hass.services.async_call(
                Platform.CLIMATE,
                SERVICE_SET_HVAC_MODE,
                {
                    ATTR_ENTITY_ID: zone.entity_id,
                    ATTR_HVAC_MODE: HVACMode.HEAT,
                },
                blocking=True,
            )

            assert mock_fcn.await_count == 1
            assert mock_fcn.await_args.args == ()
            assert mock_fcn.await_args.kwargs == {}

        # SERVICE_TURN_OFF
        with patch("evohomeasync2.zone.Zone.set_temperature") as mock_fcn:
            await hass.services.async_call(
                Platform.CLIMATE,
                SERVICE_TURN_OFF,
                {
                    ATTR_ENTITY_ID: zone.entity_id,
                },
                blocking=True,
            )

            assert mock_fcn.await_count == 1
            assert mock_fcn.await_args.args == (zone.min_temp,)
            assert mock_fcn.await_args.kwargs == {"until": None}

        # SERVICE_TURN_ON
        with patch("evohomeasync2.zone.Zone.reset_mode") as mock_fcn:
            await hass.services.async_call(
                Platform.CLIMATE,
                SERVICE_TURN_ON,
                {
                    ATTR_ENTITY_ID: zone.entity_id,
                },
                blocking=True,
            )

            assert mock_fcn.await_count == 1
            assert mock_fcn.await_args.args == ()
            assert mock_fcn.await_args.kwargs == {}

    assert results == snapshot


@pytest.mark.parametrize("install", TEST_INSTALLS)
async def test_water_heater_svcs(
    hass: HomeAssistant,
    config: dict[str, str],
    install: str,
    freezer: FrozenDateTimeFactory,
    snapshot: SnapshotAssertion,
) -> None:
    """Test water_heater services of a evohome-compatible DHW zone."""

    freezer.move_to("2024-07-10T12:00:00Z")
    results = []

    async for _ in setup_evohome(hass, config, install=install):
        if (dhw := dhw_entity(hass)) is None:
            pytest.skip("this installation has no DHW to test")

        # SERVICE_SET_AWAY_MODE: False (FollowSchedule)
        with patch("evohomeasync2.hotwater.HotWater.reset_mode") as mock_fcn:
            await hass.services.async_call(
                Platform.WATER_HEATER,
                SERVICE_SET_AWAY_MODE,
                {
                    ATTR_ENTITY_ID: dhw.entity_id,
                    ATTR_AWAY_MODE: False,
                },
                blocking=True,
            )

            assert mock_fcn.await_count == 1
            assert mock_fcn.await_args.args == ()
            assert mock_fcn.await_args.kwargs == {}

        # SERVICE_SET_AWAY_MODE: True (PermanentOverride/Off)
        with patch("evohomeasync2.hotwater.HotWater.set_off") as mock_fcn:
            await hass.services.async_call(
                Platform.WATER_HEATER,
                SERVICE_SET_AWAY_MODE,
                {
                    ATTR_ENTITY_ID: dhw.entity_id,
                    ATTR_AWAY_MODE: True,
                },
                blocking=True,
            )

            assert mock_fcn.await_count == 1
            assert mock_fcn.await_args.args == ()
            assert mock_fcn.await_args.kwargs == {}

        assert dhw.operation_list == ["auto", "on", "off"]

        # SERVICE_SET_OPERATION_MODE: "auto" (FollowSchedule)
        with patch("evohomeasync2.hotwater.HotWater.reset_mode") as mock_fcn:
            await hass.services.async_call(
                Platform.WATER_HEATER,
                SERVICE_SET_OPERATION_MODE,
                {
                    ATTR_ENTITY_ID: dhw.entity_id,
                    ATTR_OPERATION_MODE: "auto",
                },
                blocking=True,
            )

            assert mock_fcn.await_count == 1
            assert mock_fcn.await_args.args == ()
            assert mock_fcn.await_args.kwargs == {}

        # SERVICE_SET_OPERATION_MODE: "off" (TemporaryOverride/Off)
        with patch("evohomeasync2.hotwater.HotWater.set_off") as mock_fcn:
            await hass.services.async_call(
                Platform.WATER_HEATER,
                SERVICE_SET_OPERATION_MODE,
                {
                    ATTR_ENTITY_ID: dhw.entity_id,
                    ATTR_OPERATION_MODE: "off",
                },
                blocking=True,
            )

            assert mock_fcn.await_count == 1
            assert mock_fcn.await_args.args == ()
            results.append(mock_fcn.await_args.kwargs)

        # SERVICE_SET_OPERATION_MODE: "on" (TemporaryOverride/On)
        with patch("evohomeasync2.hotwater.HotWater.set_on") as mock_fcn:
            await hass.services.async_call(
                Platform.WATER_HEATER,
                SERVICE_SET_OPERATION_MODE,
                {
                    ATTR_ENTITY_ID: dhw.entity_id,
                    ATTR_OPERATION_MODE: "on",
                },
                blocking=True,
            )

            assert mock_fcn.await_count == 1
            assert mock_fcn.await_args.args == ()
            results.append(mock_fcn.await_args.kwargs)

        for service in (SERVICE_TURN_OFF, SERVICE_TURN_ON):
            with pytest.raises(HomeAssistantError):
                await hass.services.async_call(
                    Platform.WATER_HEATER,
                    service,
                    {
                        ATTR_ENTITY_ID: dhw.entity_id,
                    },
                    blocking=True,
                )

    assert results == snapshot
