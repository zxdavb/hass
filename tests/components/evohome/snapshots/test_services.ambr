# serializer version: 1
# name: test_climate_ctl_svcs[default]
  list([
    'Reset',
    'eco',
    'away',
    'home',
    'Custom',
  ])
# ---
# name: test_climate_ctl_svcs[default].1
  list([
    tuple(
      'AutoWithReset',
    ),
    tuple(
      'AutoWithEco',
    ),
    tuple(
      'Away',
    ),
    tuple(
      'DayOff',
    ),
    tuple(
      'Custom',
    ),
  ])
# ---
# name: test_climate_ctl_svcs[h032585]
  list([
  ])
# ---
# name: test_climate_ctl_svcs[h032585].1
  list([
  ])
# ---
# name: test_climate_ctl_svcs[h099625]
  list([
    'eco',
    'away',
  ])
# ---
# name: test_climate_ctl_svcs[h099625].1
  list([
    tuple(
      'AutoWithEco',
    ),
    tuple(
      'Away',
    ),
  ])
# ---
# name: test_climate_ctl_svcs[minimal]
  list([
    'Reset',
    'eco',
    'away',
    'home',
    'Custom',
  ])
# ---
# name: test_climate_ctl_svcs[minimal].1
  list([
    tuple(
      'AutoWithReset',
    ),
    tuple(
      'AutoWithEco',
    ),
    tuple(
      'Away',
    ),
    tuple(
      'DayOff',
    ),
    tuple(
      'Custom',
    ),
  ])
# ---
# name: test_climate_ctl_svcs[sys_004]
  list([
    'eco',
    'away',
  ])
# ---
# name: test_climate_ctl_svcs[sys_004].1
  list([
    tuple(
      'AutoWithEco',
    ),
    tuple(
      'Away',
    ),
  ])
# ---
# name: test_climate_zon_svcs[default]
  list([
    tuple(
      17.0,
    ),
    _Call(
      tuple(
        17.0,
      ),
      dict({
        'until': datetime.datetime(2024, 7, 10, 21, 10, tzinfo=datetime.timezone.utc),
      }),
    ),
    dict({
      'until': datetime.datetime(2024, 7, 10, 21, 10, tzinfo=datetime.timezone.utc),
    }),
  ])
# ---
# name: test_climate_zon_svcs[h032585]
  list([
    tuple(
      21.5,
    ),
    _Call(
      tuple(
        21.5,
      ),
      dict({
        'until': datetime.datetime(2024, 7, 10, 21, 10, tzinfo=datetime.timezone.utc),
      }),
    ),
    dict({
      'until': datetime.datetime(2024, 7, 10, 21, 10, tzinfo=datetime.timezone.utc),
    }),
  ])
# ---
# name: test_climate_zon_svcs[h099625]
  list([
    tuple(
      21.5,
    ),
    _Call(
      tuple(
        21.5,
      ),
      dict({
        'until': datetime.datetime(2024, 7, 10, 19, 10, tzinfo=datetime.timezone.utc),
      }),
    ),
    dict({
      'until': datetime.datetime(2024, 7, 10, 19, 10, tzinfo=datetime.timezone.utc),
    }),
  ])
# ---
# name: test_climate_zon_svcs[minimal]
  list([
    tuple(
      17.0,
    ),
    _Call(
      tuple(
        17.0,
      ),
      dict({
        'until': datetime.datetime(2024, 7, 10, 21, 10, tzinfo=datetime.timezone.utc),
      }),
    ),
    dict({
      'until': datetime.datetime(2024, 7, 10, 21, 10, tzinfo=datetime.timezone.utc),
    }),
  ])
# ---
# name: test_climate_zon_svcs[sys_004]
  list([
    tuple(
      15.0,
    ),
    _Call(
      tuple(
        15.0,
      ),
      dict({
        'until': datetime.datetime(2024, 7, 10, 20, 10, tzinfo=datetime.timezone.utc),
      }),
    ),
    dict({
      'until': None,
    }),
  ])
# ---
# name: test_evohome_ctl_svcs[default]
  list([
    'refresh_system',
    'reset_system',
    'set_system_mode',
    'clear_zone_override',
    'set_zone_override',
  ])
# ---
# name: test_evohome_ctl_svcs[default].1
  list([
    dict({
      'until': HAFakeDatetime(2024, 7, 31, 7, 0, tzinfo=datetime.timezone.utc),
    }),
  ])
# ---
# name: test_evohome_ctl_svcs[h032585]
  list([
    'refresh_system',
    'set_system_mode',
    'clear_zone_override',
    'set_zone_override',
  ])
# ---
# name: test_evohome_ctl_svcs[h099625]
  list([
    'refresh_system',
    'set_system_mode',
    'clear_zone_override',
    'set_zone_override',
  ])
# ---
# name: test_evohome_ctl_svcs[minimal]
  list([
    'refresh_system',
    'reset_system',
    'set_system_mode',
    'clear_zone_override',
    'set_zone_override',
  ])
# ---
# name: test_evohome_ctl_svcs[minimal].1
  list([
    dict({
      'until': HAFakeDatetime(2024, 7, 31, 7, 0, tzinfo=datetime.timezone.utc),
    }),
  ])
# ---
# name: test_evohome_ctl_svcs[sys_004]
  list([
    'refresh_system',
    'set_system_mode',
    'clear_zone_override',
    'set_zone_override',
  ])
# ---
# name: test_evohome_zon_svcs[default]
  list([
    dict({
      'until': datetime.datetime(2024, 7, 10, 21, 10, tzinfo=datetime.timezone.utc),
    }),
  ])
# ---
# name: test_evohome_zon_svcs[h032585]
  list([
    dict({
      'until': datetime.datetime(2024, 7, 10, 21, 10, tzinfo=datetime.timezone.utc),
    }),
  ])
# ---
# name: test_evohome_zon_svcs[h099625]
  list([
    dict({
      'until': datetime.datetime(2024, 7, 10, 19, 10, tzinfo=datetime.timezone.utc),
    }),
  ])
# ---
# name: test_evohome_zon_svcs[minimal]
  list([
    dict({
      'until': datetime.datetime(2024, 7, 10, 21, 10, tzinfo=datetime.timezone.utc),
    }),
  ])
# ---
# name: test_evohome_zon_svcs[sys_004]
  list([
    dict({
      'until': datetime.datetime(2024, 7, 10, 20, 10, tzinfo=datetime.timezone.utc),
    }),
  ])
# ---
# name: test_water_heater_svcs[default]
  list([
    dict({
      'until': datetime.datetime(2024, 7, 10, 12, 0, tzinfo=datetime.timezone.utc),
    }),
    dict({
      'until': datetime.datetime(2024, 7, 10, 12, 0, tzinfo=datetime.timezone.utc),
    }),
  ])
# ---
